src/gbinder_fmq.c

```diff
@@ -650,77 +600,60 @@ gbinder_fmq_write(
             }
 
             gbinder_fmq_end_write(self, items);
-            ret = TRUE;
+            return TRUE;
         }
     }
-
-    return ret;
+    return FALSE;
 }
 
 int
 gbinder_fmq_wait_timeout(
     GBinderFmq* self,
     guint32 bit_mask,
     guint32* state,
-    guint timeout_ms)
+    int timeout_ms)
 {
-    int ret = 0;
-    if (G_LIKELY(self) && G_LIKELY(state)) {
-        /* Event flag is not configured */
-        if (self->event_flag_ptr == NULL) {
-            ret = -ENOSYS;
-        } else if (bit_mask == 0 || state == NULL) {
-            ret = -EINVAL;
+    if (G_UNLIKELY(!state) || G_UNLIKELY(!self)) {
+        return (-EINVAL);
+    } else if (!self->event_flag_ptr) {
+        return (-ENOSYS);
+    } else if (!bit_mask) {
+        return (-EINVAL);
+    } else {
+        guint32 old_value = atomic_fetch_and(self->event_flag_ptr, ~bit_mask);
+        guint32 set_bits = old_value & bit_mask;
+
+        /* Check if any of the bits was already set */
+        if (set_bits != 0) {
+            *state = set_bits;
+            return 0;
+        } else if (!timeout_ms) {
+            return (-EAGAIN);
         } else {
-            guint32 old_value = atomic_fetch_and(self->event_flag_ptr,
-                ~bit_mask);
-            guint32 set_bits = old_value & bit_mask;
+            int ret;
+
+            if (timeout_ms > 0) {
+                struct timespec wait_time;
 
-            /* Check if any of the bits was already set */
-            if (set_bits != 0) {
-                *state = set_bits;
+                clock_gettime(CLOCK_MONOTONIC, &wait_time);
+                wait_time.tv_sec += timeout_ms / 1000;
+                wait_time.tv_nsec += (timeout_ms % 1000) * 1000000;
```

Comment:
(timeout_ms % 1000) can't be greater than 999 meaning that tv_nsec is limited by 999000000 which is smaller than 999999999.

Reply:
Right, I missed that. The variable name confused me (wait_time sounds like a period rather than absolute time). Good catch. Will fix.

That must be why I was sometimes getting EINVAL instead of ETIMETOUT in unit tests. Will also roll back the unit test changes which I made because of that.

Reply:
Fixed. And check for ETIMETOUT in unit_fmq.c is back, too.